Deploy Docs (#3)
IntuitiveWebSolutions/gen3-template: build/README.md

<!-- template-doc -->

# Deployment How-to

## Setting up Cloudformation

> AWS Cloudformation can create all the resources you need to deploy your project using the [dev-resources.yaml template](https://github.com/IntuitiveWebSolutions/gen3-template/blob/aaronsaloff/deploy_docs/dev-resources.yaml)

Reviewer: Relative path

```suggestion
AWS Cloudformation can create all the resources you need to deploy your project using the [dev-resources.yaml template](/dev-resources.yaml)
```

**Make sure to look at all the comments in that file and customize it for your needs.**

To create your Cloudformation stack use the [create-stack AWS-CLI command](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/create-stack.html)

You can use `aws-okta` for this (https://britecore.slack.com/archives/C9F2CCWKE/p1557510366125400)

Example:
```
aws-okta exec service-lead  -- aws cloudformation create-stack --stack-name <Whatever-your-naming-it> --template-body file://dev-resources.yaml --capabilities CAPABILITY_IAM --role-arn arn:aws:iam::444556543218:role/BC-CloudFormation-Role
```

> Note: If you use `aws-okta exec sandbox` you can create your project in the sandbox account and have more flexibility while you're hashing things out. Just make sure to delete the project when you move out of the sandbox.

Reviewer: ```suggestion
**Note:** If you use `aws-okta exec sandbox` you can create your project in the sandbox account and have more flexibility while you're hashing things out. Just make sure to delete the project when you move out of the sandbox.
```

## Using the ui.britecore.com CDN

> Since the [Design System repo's Cloudformation template](https://github.com/IntuitiveWebSolutions/BC-Design-System/blob/master/dev-resources.yaml) creates the CDN the uses ui.britecore.com we will need to add our newly created bucket as an origin in the Design Systems CDN and add the cache behavior that will route our requests accordingly.

Reviewer 2: Maybe we should add a note here that  communicates that we will be moving that configuration to a centralized repo.

Author: Yes definitely! I'll add that.

> Since the [Design System repo's Cloudformation template](https://github.com/IntuitiveWebSolutions/BC-Design-System/blob/master/dev-resources.yaml) creates the CDN the uses ui.britecore.com we will need to add our newly created bucket as an origin in the Design Systems CDN and add the cache behavior that will route our requests accordingly.

Reviewer: ```suggestion
Since the [Design System repo's CloudFormation template](https://github.com/IntuitiveWebSolutions/BC-Design-System/blob/master/dev-resources.yaml) creates the CDN that uses ui.britecore.com, we will need to add our newly created bucket as an origin in the Design System's CDN and add the cache behavior that will route our requests accordingly.
```

In the CDN resource under `Origins` we will need to add our bucket. Here is an example of the one from the Accounts Receivable page:

```
Origins:
  - Id:
      Ref: DocsWebsiteBucket
    DomainName:
      Fn::GetAtt: [DocsWebsiteBucket, DomainName]
    S3OriginConfig:
      OriginAccessIdentity:
        Fn::Join:
          - '/'
          - - origin-access-identity
            - cloudfront
            - Ref: OriginAccessIdentity

>   - Id: ui.britecore.com/pages/accounts-receivable

Reviewer 2: Should we add placeholders for these values?

Reviewer: +1

Reviewer 2: ```suggestion
  - Id: ui.britecore.com/pages/{your-page-name}
```

>     DomainName: gen3-accounts-receivable-development-assetsbucket-vtfcgxiuh5is.s3.amazonaws.com

Reviewer 2: Same as above ☝️

Reviewer 2: ```suggestion
    DomainName: gen3-{your-page-name}-development-assetsbucket-vtfcgxiuh5is.s3.amazonaws.com
```

    S3OriginConfig:
      OriginAccessIdentity:
        Fn::Join:
          - '/'
          - - origin-access-identity
            - cloudfront
            - Ref: OriginAccessIdentity
```

We then need to add a cache behavior to the CDN that will route our requests to the appropiate bucket:

```
CacheBehaviors:

>   - TargetOriginId: ui.britecore.com/pages/accounts-receivable

Reviewer 2: Same as above

Reviewer 2: ```suggestion
  - TargetOriginId: ui.britecore.com/pages/{your-page-name}
```

>     PathPattern: 'pages/accounts-receivable/*'

Reviewer 2: Same as above

Reviewer 2: ```suggestion
    PathPattern: 'pages/{your-page-name}/*'
```

    AllowedMethods: [GET, HEAD, OPTIONS]
    CachedMethods: [GET, HEAD, OPTIONS]
    Compress: true
    DefaultTTL: 86400
    ForwardedValues:
      QueryString: false
    MaxTTL: 31536000
    MinTTL: 0
    SmoothStreaming: false
    ViewerProtocolPolicy: redirect-to-https
```

> To avoid route collisions, we've designated the `pages` namespace to route to pages outside of the Design System's documentation website. Make sure your S3 bucket has the correct directory structure. We have the Accounts Receivable page at `/pages/accounts-receivable/`

Reviewer 2: ```suggestion
To avoid route collisions, we've designated the `pages` namespace to route to conversion application assets. Since assets from all conversions will live at this name space, it is important that you deploy your assets to to a sub-route under `pages`. For example, the Accounts Receivable page is deployed to `/pages/accounts-receivable/`. 

⚠️ **Note**: the router expects to find the directory structure specified by your route. For the example above, the accounts receivable assets must be deployed to an `accounts-receivable` directory that lives within a `pages` directory. The `pages` directory, in turn, sits at the root.
```

## Versioning

We've configured webpack to no longer add chunk hashes in file names. Instead, we are versioning using folder structure. This enables us to make explicit PRs when making changes to Gen2 and also allows us to easily revert if there are issues with a release.

> To set this up make sure to edit both [`publish.sh`](https://github.com/IntuitiveWebSolutions/gen3-template/blob/master/build/bin/publish.sh) and [`vue.config.js`](https://github.com/IntuitiveWebSolutions/gen3-template/blob/master/vue.config.js) where commented to make sure your S3 bucket's directory structure is correct. Again, the parent directory must be `pages` and then there should be a subdirectory name that you can store the version folders in.

Reviewer: Relative paths

```suggestion
To set this up make sure to edit both [`publish.sh`](./bin/publish.sh) and [`vue.config.js`](/vue.config.js) where commented to make sure your S3 bucket's directory structure is correct. Again, the parent directory must be `pages` and then there should be a subdirectory name that you can store the version folders in.
```

> If you've followed the steps above, whenever you change the version in `package.json` it will create a new directory in your S3 bucket with that verison's number and store the built project under that directory. Here's what it would look like:

Reviewer: ```suggestion
If you've followed the steps above, whenever you change the version in `package.json` it will create a new directory in your S3 bucket with that version's number and store the built project under that directory. Here's what it would look like:
```

```
pages
  |
  └ <your-namespace>
    |
    └ 0.5.0
    |
    └ 0.6.0
      css
      fonts
      img
      js
      index.html
```

> You're latest `app.js` file can now be accessed by visiting `http://ui.britecore.com/pages/<your-namespace>/0.6.0/js/app.js`

Reviewer: ```suggestion
Your latest `app.js` file can now be accessed by visiting `https://ui.britecore.com/pages/<your-namespace>/0.6.0/js/app.js`
```

## Plugging the project into a Gen2 page

> The Vue instance created in this project can be injected onto any Gen2 page by including the correct CSS and Javascript files from the CDN you deployed in the above steps. Make sure you include a div with the id specified in `main.js` when creating the Vue instance.

Reviewer: ```suggestion
The Vue instance created in this project can be injected into any Gen2 page by including the correct CSS and Javascript files from the CDN you deployed in the above steps. Make sure you include a div with the id specified in [`main.js`](/src/main.js) when creating the Vue instance.
```

Author: Done

Example:

```

> <link href="http://ui.britecore.com/pages/accounts-receivable/0.6.0/css/chunk-vendors.css" rel="stylesheet">

Reviewer: ```suggestion
<link href="https://ui.britecore.com/pages/accounts-receivable/0.6.0/css/chunk-vendors.css" rel="stylesheet">
```

> <link href="http://ui.britecore.com/pages/accounts-receivable/0.6.0/css/app.css" rel="stylesheet">

Reviewer: Same as above

Author: Done

<div id="ar-vue-instance"></div>

> <script src="http://ui.britecore.com/pages/accounts-receivable/0.6.0/js/chunk-vendors.js"></script>

Reviewer: Same as above (https)

> <script src="http://ui.britecore.com/pages/accounts-receivable/0.6.0/js/app.js"></script>

Reviewer: Same as above (https)

Author: Done

```

You can also test it locally by doing the following:

1. Run a local instance of Gen2
2. Build this project with `yarn build`
3. Install a command line http server: 'yarn global add http-server'
4. In this projects directory run `http-server dist -p 8081 --cors='*'`
5. Replace the asset URLs in Gen2 with the local one like so:

```
<link href="http://localhost:8081/css/chunk-vendors.css" rel="stylesheet">
<link href="http://localhost:8081/css/app.css" rel="stylesheet">
<div id="ar-vue-instance"></div>
<script src="http://localhost:8081/js/chunk-vendors.js"></script>
<script src="http://localhost:8081/js/app.js"></script>
```

## Feature flag

To put your page behind a setting in Gen2 you'll need to add a postfix and a line for `initial.sql`.

> Here are examples of a [postfix script](https://github.com/IntuitiveWebSolutions/BriteCore/pull/12075/files#diff-88fc50db6a463f4d1dc85dd6975f0f16) and a [entry added to intial.sql](https://github.com/IntuitiveWebSolutions/BriteCore/pull/12075/files#diff-a749d68281ea0572e573f16e526357a4)

Reviewer: ```suggestion
Here are examples of a [postfix script](https://github.com/IntuitiveWebSolutions/BriteCore/pull/12075/files#diff-88fc50db6a463f4d1dc85dd6975f0f16) and an [entry added to intial.sql](https://github.com/IntuitiveWebSolutions/BriteCore/pull/12075/files#diff-a749d68281ea0572e573f16e526357a4)
```

> Note: You can generate a UUID online that can be used for those.

Reviewer: ```suggestion
**Note:** You can generate a UUID online that can be used for those.
```

> Once you have the setting in Gen2, you need to run `bin/postfix.py` inside your local Gen2 webserver and you should be able to acces the setting under `Settings > Advanced`.

Reviewer: ```suggestion
Once you have the setting in Gen2, you need to run `./bin/postfix.py` inside your local Gen2 container and you should be able to access the setting under `Settings > Advanced` in the browser.
```

Might be a good idea to mention how to get inside the container and the usage of `activate_britecore`

We can now use the setting in our Mako templates in Gen2. See here for an example. https://github.com/IntuitiveWebSolutions/BriteCore/blob/master/templates/britecore/policies/accountsReceivable.html

> ### "Switch back" functionality

Reviewer 2: ```suggestion
### "Back to legacy page" functionality
```

> You can add a "Back to legacy page" button by using a query string in your URL. The above example only shows the new Accounts Receivable page if the setting is on AND the URL does not have the `legacy=1` query string in it.

Reviewer 2: ```suggestion
As you're rolling out your new page to clients, some user will occasionally want to turn back to the legacy version of the page. The example provided in the last section, demonstrates how this feature can be implemented by adding query parameters to the page URL. The implementation boils down to reading a query parameter, say, `legacy=1` on the server, and wrapping the new version of the page on a conditional that only renders if the feature flag is set and `legacy != 1`. By evaluating this condition at the server render step, only the bytes for the chosen page version are sent back. 
```

## Versioning

We've configured webpack to no longer add chunk hashes in file names. Instead, we are versioning using folder structure. When a new version is deployed it is placed in a folder named after the new version. This allows for easy reversions, which can be made by simply changing the folder name in the request path for the project files.

To set this up make sure to edit both [`publish.sh`](./bin/publish.sh) and [`vue.config.js`](/vue.config.js) where commented to make sure your S3 bucket's directory structure is correct. Again, the parent directory must be `pages` and then there should be a subdirectory name that you can store the version folders in.

If you've followed the steps above, whenever you change the version in `package.json` it will create a new directory in your S3 bucket with that version's number and store the built project under that directory. Here's what it would look like:

```
pages
  |

>   └ <your-namespace>

Reviewer 2: ```suggestion
  └ <your-page-name>
```
